### PR TITLE
Print op

### PR DESCRIPTION
**Pull request by snf**
This PR is adding the print_op that will behave as the identity function but will print the tensor value when evaluated.
I decided to name it print_op because print is a reserved word in python but I'm open to suggestions.

### CODE REVIEW

**Comment by snf** [comment]
Just squashed the commits into one.

**Comment by EderSantana** [comment]
@snf this is really useful specially for theano
also could you add the choice of printing variable attributes to do things such as:
```python
x = theano.printing.Print('[Debug] X shape is: ', attrs=("shape",))(x)
```
maybe passing an `attributes` input to `print_op`?

also, maybe it could just be called `print`

**Comment by snf** [comment]
@EderSantana yeah, I tried to keep it as simple as possible with minimal compatibility functionality.
The reason it's named `print_op` instead of `print` is because it's a keyword in python and I'm not sure it's completely safe to use it.

**Comment by snf** [comment]
@EderSantana btw, the shape is "static", I'm not sure how useful it is to print it on evaluation.

**Comment by EderSantana** [comment]
you are right we can't name it just `print`

**Comment by snf** [comment]
I was just thinking and maybe we can make this like:

    def print_op(x, message='', shape=False):
      attrs = ['__str__']
      if shape:
        attrs.append('shape')
      op = Print(message, attrs)
      return op(x)

And ignore the shape argument on tensorflow?, I'm not sure how to solve this (I know there could be many other attrs wanted), I would rather leave it as a simple print.

**Comment by farizrahman4u** [comment]
This will be the most useful op in Keras. Keep it simple imo, just tensor value and shape. Instead of complicating things with arguments, just write 2 ops, `print_op` and `print_shape_op`.

**Comment by farizrahman4u** [comment]
Also add corresponding layers in core.py.

**Comment by fchollet** [comment]
I don't think it's necessary to have a layer for `print`. It can always be used with a `Lambda` or inside a custom layer.

I'd rename it to `print_tensor` (not a fan of the `op` suffix for ops).

**Comment by snf** [comment]
Ok, renamed it to print_tensor

**Comment by farizrahman4u** [comment]
> I don't think it's necessary to have a layer for print

Is that open to vote?

> I'd rename it to print_tensor (not a fan of the op suffix for ops).

May I suggest `print_value` and `print_shape`?

**Comment by snf** [comment]
@farizrahman4u I used this to debug one of my network implementations and it becomes really verbose so leaving this enabled was a no-no for more than a few mini-batches. I don't know if a layer would be good though.
About the `print_value` and `print_shape`, I think we already have `K.shape.eval`. I don't know if it's really useful to being able to print the shape (maybe I just didn't come across a case yet).

**Comment by farizrahman4u** [comment]
Cool! Its upto you and Francois. Layers will be ridiculously verbose unless redirected to file stream.

**Comment by fchollet** [comment]
LGTM

**Comment by ParikhKadam** [comment]
> May I suggest print_value and print_shape?
@farizrahman4u Yes, we need it. I too need the support for print_shape.

> About the print_value and print_shape, I think we already have K.shape.eval. I don't know if it's really useful to being able to print the shape (maybe I just didn't come across a case yet).
@snf Can you share a code snippet to do this. I need to print the shape of tensors at time of training in order to debug my model.

I tried the implementation using a Lambda layer but it throws me an error saying that it needs the shape of tensor at the time of compiling the layer instead of training time. Here's the method I used to implement it:

Code:
```
from keras import backend as K
def func(x):
    return K.eval(K.shape(x))(x, 'question_embedding')
question_embedding = Lambda(func)(question_embedding)
```

Error:
```
InvalidArgumentError                      Traceback (most recent call last)
~/eassistant/env/lib/python3.6/site-packages/tensorflow/python/client/session.py in _do_call(self, fn, *args)
   1333     try:
-> 1334       return fn(*args)
   1335     except errors.OpError as e:

~/eassistant/env/lib/python3.6/site-packages/tensorflow/python/client/session.py in _run_fn(feed_dict, fetch_list, target_list, options, run_metadata)
   1318       return self._call_tf_sessionrun(
-> 1319           options, feed_dict, fetch_list, target_list, run_metadata)
   1320 

~/eassistant/env/lib/python3.6/site-packages/tensorflow/python/client/session.py in _call_tf_sessionrun(self, options, feed_dict, fetch_list, target_list, run_metadata)
   1406         self._session, options, feed_dict, fetch_list, target_list,
-> 1407         run_metadata)
   1408 

InvalidArgumentError: You must feed a value for placeholder tensor 'question_input_1' with dtype float and shape [?,?,400]
	 [[{{node question_input_1}} = Placeholder[dtype=DT_FLOAT, shape=[?,?,400], _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

During handling of the above exception, another exception occurred:

InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-31-31d500da1d83> in <module>
      2 def func(x):
      3     return K.eval(K.shape(x))(x, 'question_embedding')
----> 4 question_embedding = Lambda(func)(question_embedding)

~/eassistant/env/lib/python3.6/site-packages/keras/engine/base_layer.py in __call__(self, inputs, **kwargs)
    455             # Actually call the layer,
    456             # collecting output(s), mask(s), and shape(s).
--> 457             output = self.call(inputs, **kwargs)
    458             output_mask = self.compute_mask(inputs, previous_mask)
    459 

~/eassistant/env/lib/python3.6/site-packages/keras/layers/core.py in call(self, inputs, mask)
    685         if has_arg(self.function, 'mask'):
    686             arguments['mask'] = mask
--> 687         return self.function(inputs, **arguments)
    688 
    689     def compute_mask(self, inputs, mask=None):

<ipython-input-31-31d500da1d83> in func(x)
      1 from keras import backend as K
      2 def func(x):
----> 3     return K.eval(K.shape(x))(x, 'question_embedding')
      4 question_embedding = Lambda(func)(question_embedding)

~/eassistant/env/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py in eval(x)
    671     ```
    672     """
--> 673     return to_dense(x).eval(session=get_session())
    674 
    675 

~/eassistant/env/lib/python3.6/site-packages/tensorflow/python/framework/ops.py in eval(self, feed_dict, session)
    711 
    712     """
--> 713     return _eval_using_default_session(self, feed_dict, self.graph, session)
    714 
    715 

~/eassistant/env/lib/python3.6/site-packages/tensorflow/python/framework/ops.py in _eval_using_default_session(tensors, feed_dict, graph, session)
   5155                        "the tensor's graph is different from the session's "
   5156                        "graph.")
-> 5157   return session.run(tensors, feed_dict)
   5158 
   5159 

~/eassistant/env/lib/python3.6/site-packages/tensorflow/python/client/session.py in run(self, fetches, feed_dict, options, run_metadata)
    927     try:
    928       result = self._run(None, fetches, feed_dict, options_ptr,
--> 929                          run_metadata_ptr)
    930       if run_metadata:
    931         proto_data = tf_session.TF_GetBuffer(run_metadata_ptr)

~/eassistant/env/lib/python3.6/site-packages/tensorflow/python/client/session.py in _run(self, handle, fetches, feed_dict, options, run_metadata)
   1150     if final_fetches or final_targets or (handle and feed_dict_tensor):
   1151       results = self._do_run(handle, final_targets, final_fetches,
-> 1152                              feed_dict_tensor, options, run_metadata)
   1153     else:
   1154       results = []

~/eassistant/env/lib/python3.6/site-packages/tensorflow/python/client/session.py in _do_run(self, handle, target_list, fetch_list, feed_dict, options, run_metadata)
   1326     if handle is None:
   1327       return self._do_call(_run_fn, feeds, fetches, targets, options,
-> 1328                            run_metadata)
   1329     else:
   1330       return self._do_call(_prun_fn, handle, feeds, fetches)

~/eassistant/env/lib/python3.6/site-packages/tensorflow/python/client/session.py in _do_call(self, fn, *args)
   1346           pass
   1347       message = error_interpolation.interpolate(message, self._graph)
-> 1348       raise type(e)(node_def, op, message)
   1349 
   1350   def _extend_graph(self):

InvalidArgumentError: You must feed a value for placeholder tensor 'question_input_1' with dtype float and shape [?,?,400]
	 [[node question_input_1 (defined at /home/cedept/eassistant/env/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:517)  = Placeholder[dtype=DT_FLOAT, shape=[?,?,400], _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

Caused by op 'question_input_1', defined at:
  File "/usr/local/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/ipykernel_launcher.py", line 16, in <module>
    app.launch_new_instance()
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/traitlets/config/application.py", line 658, in launch_instance
    app.start()
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/ipykernel/kernelapp.py", line 505, in start
    self.io_loop.start()
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tornado/platform/asyncio.py", line 132, in start
    self.asyncio_loop.run_forever()
  File "/usr/local/lib/python3.6/asyncio/base_events.py", line 421, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.6/asyncio/base_events.py", line 1426, in _run_once
    handle._run()
  File "/usr/local/lib/python3.6/asyncio/events.py", line 127, in _run
    self._callback(*self._args)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tornado/ioloop.py", line 758, in _run_callback
    ret = callback()
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tornado/stack_context.py", line 300, in null_wrapper
    return fn(*args, **kwargs)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tornado/gen.py", line 1233, in inner
    self.run()
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tornado/gen.py", line 1147, in run
    yielded = self.gen.send(value)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/ipykernel/kernelbase.py", line 357, in process_one
    yield gen.maybe_future(dispatch(*args))
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tornado/gen.py", line 326, in wrapper
    yielded = next(result)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/ipykernel/kernelbase.py", line 267, in dispatch_shell
    yield gen.maybe_future(handler(stream, idents, msg))
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tornado/gen.py", line 326, in wrapper
    yielded = next(result)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/ipykernel/kernelbase.py", line 534, in execute_request
    user_expressions, allow_stdin,
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tornado/gen.py", line 326, in wrapper
    yielded = next(result)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/ipykernel/ipkernel.py", line 294, in do_execute
    res = shell.run_cell(code, store_history=store_history, silent=silent)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/ipykernel/zmqshell.py", line 536, in run_cell
    return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2819, in run_cell
    raw_cell, store_history, silent, shell_futures)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2845, in _run_cell
    return runner(coro)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/IPython/core/async_helpers.py", line 67, in _pseudo_sync_runner
    coro.send(None)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3020, in run_cell_async
    interactivity=interactivity, compiler=compiler, result=result)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3185, in run_ast_nodes
    if (yield from self.run_code(code, result)):
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3267, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-27-95e282f86396>", line 1, in <module>
    question_input = Input(shape=(None, emdim), dtype='float32', name="question_input")
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/keras/engine/input_layer.py", line 178, in Input
    input_tensor=tensor)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/keras/legacy/interfaces.py", line 91, in wrapper
    return func(*args, **kwargs)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/keras/engine/input_layer.py", line 87, in __init__
    name=self.name)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 517, in placeholder
    x = tf.placeholder(dtype, shape=shape, name=name)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tensorflow/python/ops/array_ops.py", line 1747, in placeholder
    return gen_array_ops.placeholder(dtype=dtype, shape=shape, name=name)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 5206, in placeholder
    "Placeholder", dtype=dtype, shape=shape, name=name)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/home/cedept/eassistant/env/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

InvalidArgumentError (see above for traceback): You must feed a value for placeholder tensor 'question_input_1' with dtype float and shape [?,?,400]
	 [[node question_input_1 (defined at /home/cedept/eassistant/env/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:517)  = Placeholder[dtype=DT_FLOAT, shape=[?,?,400], _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]
```

In short, it says that K.eval() needs its actual parameters to evaluate it right now... And it won't wait until the time we call model.fit(). Any way I can get tensor shapes at run time instead of compile time?
